app/models/concerns/galtersufia/generic_file/header_lookup.rb

```diff
@@ -0,0 +1,76 @@
+module Galtersufia
+  module GenericFile
+    module HeaderLookup
+      extend ActiveSupport::Concern
+
+      BASE_SPARQL_MESH_URI ="https://id.nlm.nih.gov/mesh/sparql?format=JSON&limit=10&inference=true&query=PREFIX%20rdfs"\
+        "%3A%20%3Chttp%3A%2F%2Fwww.w3.org%2F2000%2F01%2Frdf-schema%23%3E%0D%0APREFIX%20meshv%3A%20%3Chttp%3A%2F%2Fid.nlm."\
+        "nih.gov%2Fmesh%2Fvocab%23%3E%0D%0APREFIX%20mesh2018%3A%20%3Chttp%3A%2F%2Fid.nlm.nih.gov%2Fmesh%3E%0D%0A%0D%0ASELECT"\
+        "%20%3Fd%20%3FdName%0D%0AFROM%20%3Chttp%3A%2F%2Fid.nlm.nih.gov%2Fmesh%3E%0D%0AWHERE%20%7B%0D%0A%20%20%3Fd%20a%20"\
+        "meshv%3ADescriptor%20.%0D%0A%20%20%3Fd%20rdfs%3Alabel%20%3FdName%0D%0A%20%20FILTER(REGEX(%3FdName%2C%27"
+      END_SPARQL_MESH_URI = "%27%2C%20%27i%27))%20%0D%0A%7D%20%0D%0AORDER%20BY%20%3Fd%20%0D%0A"
+      LCNAF_BASE_URI = "http://id.loc.gov/authorities/names/suggest/?q="
+
+      # TODO figure out how to test this...
+      # return PID for provided mesh_header using SPARQL query
+      def mesh_term_lookup(mesh_term="")
+        return unless mesh_term.present?
+
+        result = memoized_mesh_lookups[mesh_term]
+        return result if result.present?
+
+        query_result  = HTTParty.get(BASE_SPARQL_MESH_URI + mesh_term + END_SPARQL_MESH_URI)
+        hits = JSON.parse(query_result)
+
+        if hits = hits.try(:[], "results").try(:[], "bindings")
+          result = pid_from_mesh_hits(hits)
+          memoized_mesh_lookups[mesh_term] = result
+          result
+        else
+          ["Error Searching"]
+        end
+      end
+
+      def lcsh_term_lookup(lcsh_term="")
+        return unless lcsh_term.present?
+
+        result = memoized_lcsh_lookups[lcsh_term]
+        return result if result.present?
+
+        # perform search with lcsh_term's whitespace replaced with '*' character
+        query_result = HTTParty.get(LCNAF_BASE_URI + "*#{lcsh_term.gsub(/\s/,'*')}*")
+        hits = JSON.parse(query_result)
+
+        if hits = hits.try(:[], 1)
+          result = pid_from_lcsh_hits(hits)
+          memoized_lcsh_lookups[lcsh_term] = result
+        else
+          ["Error Searching"]
+        end
+      end
+
+      private
+
+      def pid_from_mesh_hits(hits)
+        # TODO identify situations where there will be multiple "hits" - pids that match the search
+        # result = hits.map{|x| x["d"]["value"].split('/').last}
+        # below is solution for singlular "hit"
+        hits.shift["d"]["value"].split('/').last
```

Comment:
I really hate this, but I'm not really sure how else you can go about accessing such deeply nested values. At least it's wrapped in a function I guess.